### DISCUSSION TITLE
Using lsp docker with multiple projects and language-servers

### DISCUSSION THREAD

**Discussion by ratking82** [emacs-lsp/lsp-docker#66] [Q&A]
Hello @factyy, 

I tried using `lsp-docker` with persistent config based on YAML but ran into a few issue. Could you please help me resolve them ?  
I will try and explain the issues in details below.

I have two projects A & B. Project A consists of C/C++, Python and shell scripts. While Project B has mainly C code. Project A uses a container based dev environment. I deployed `clangd` into the dev container of Project A such that clang can find all lib sources etc and used the manual langserver deployment guide in the lsp-docker repo. It worked without an issue. The problem started when I tried to configure the container to also have `pyright`. It seems I cannot create a YAML based config which can create separate containers from the same image file. 
The second issue is with Project B. Project B uses the `emacslsp/lsp-docker-langservers` image. But every time I open any file inside the project the container for Project A starts up. As Project A has different path mappings than Project B, lsp fails to start the server.  Also as Project A was configured with clangd which is not available in the `emacslsp/lsp-docker-langservers` image the image could not start up. I added `clangd` to the image and retried and that is when I noticed the above issue. 

Thank you for your help in advance.

**Comment by factyy** [comment]
@ratking82 , thank you for bringing this up! 

Answering the first question : when I was adding support for persistent configuration the idea was to create a single image or container for the project, so apparently I haven't added any features to support this case. To be clear though: you want to use a single image with two different entry points? And allow selecting the right one based on a certain criteria (like having the right major mode), right?

And the second question: I remember adding a path tests to the image / container selection. Are paths of Project A and B somehow related?

**Reply by ratking82** [reply]
Hello @factyy ,

Thank you replying to my query. 
1. Yes, I would like to create a container image containing all libraries and tools required for development on the current code base. Thus, it will also have all lang-servers which are needed and then have the correct lang server started in a container depending on the major mode. I do not know if two lang servers can be started in the same container. Currently, the `emacslsp/lsp-docker-langservers` image does start a container with the correct lang-server when editing a file but it fails to respect path-mappings on per project basis.
2. The common parts of the two projects is that they both reside in my home directory under a directory named `code`. Thus `$HOME/code/project_A` and `$HOME/code/project_B`.

**Reply by factyy** [reply]
Ok, now I fully understand your first request: I'll try to find a way to get it done, but I'm really not sure about whether it is possible. But if it is not possible, maybe it is a good idea to have such functionality in the future :)

For the second part: it's a bit weird, it has to check the project path (checking via `lsp-` functions, particularly `lsp-workspace-root`) as part of the decision making process. 

Sadly right now I really don't have enough time, so your help is really appreciated :) You can start by checking the `lsp-docker-create-activation-function-by-project-dir` invocations - debug whether the activation function is called properly.

**Reply by ratking82** [reply]
Okay I will give it a shot!
